witness/golang/omniwitness/usbarmory/internal/storage/slots/slots.go

```diff
@@ -116,6 +117,14 @@ func (p *Partition) Open(slot uint) (*Slot, error) {
 		return nil, fmt.Errorf("invalid slot %d (partition has %d slots)", slot, l)
 	}
 	s := &p.slots[slot]
+	if s.journal == nil {
+		glog.V(2).Infof("Opening journal for slot %d", slot)
+		j, err := OpenJournal(p.dev, s.start, s.length, p.sha256)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open journal for slot %d: %v", slot, err)
+		}
+		s.journal = j
+	}
```

Comment:
I _think_ this should be an `EnsureInit` method on `Slot`, and it should be guarded by the lock there. As this stands, unless we know that this `Open` is only called from a single thread, multiple journals could be opened for the same slot. Obviously only the last one wins, but I suspect it's a source of hard to find bugs and not how we reason the initialization should work.

Reply:
Yeah, good spot - this is better, ta.